errors.go

```diff
@@ -346,3 +351,70 @@ func (e *wrappedErr) WithMeta(key string, val string) Error {
 	}
 }
 func (e *wrappedErr) Cause() error { return e.cause }
+
+// WriteError writes an HTTP response with a valid Twirp error format (code, msg, meta).
+// Useful outside of the Twirp server (e.g. http middleware).
+// If err is not a twirp.Error, it will get wrapped with twirp.InternalErrorWith(err)
+func WriteError(resp http.ResponseWriter, err error) {
+	writeError(resp, err)
+}
+
+// writeError writes Twirp errors in the response.
+func writeError(resp http.ResponseWriter, err error) {
+	// Non-twirp errors are wrapped as Internal (default)
+	twerr, ok := err.(Error)
+	if !ok {
+		twerr = InternalErrorWith(err)
+	}
+
+	statusCode := ServerHTTPStatusFromErrorCode(twerr.Code())
+	respBody := marshalErrorToJSON(twerr)
+
+	resp.Header().Set("Content-Type", "application/json") // Error responses are always JSON
+	resp.Header().Set("Content-Length", strconv.Itoa(len(respBody)))
+	resp.WriteHeader(statusCode) // set HTTP status code and send response
+
+	_, writeErr := resp.Write(respBody)
+	if writeErr != nil {
+		// We have two options here. We could log the error, or just silently
+		// ignore the error.
+		//
+		// Logging is unacceptable because we don't have a user-controlled
+		// logger; writing out to stderr without permission is too rude.
+		//
+		// Silently ignoring the error is our least-bad option. It's highly
+		// likely that the connection is broken and the original 'err' says
+		// so anyway.
```

Comment:
or return the error to the caller and let them log/ignore?

Reply:
Ah! You're completely right that this is a very different situation from the generated code. This should most certainly be able to return an error.

It's unfortunate that this would be a breaking change and require a major version bump. This is why public API is hard. Ouch.

Reply:
Indeed this could return an error, and probably should within this pkg.

But on the flip-side, it would no longer be a drop-in replacement to the generated version of `WriteError`.

In our case we were importing stub packages in various places outside the twirp stack, so it made sense to have it this way.

But it wouldn't be too hard to refactor with the returned error (we'd probably drop it anyways).

Maybe in a v6 we could add a returned error, or in v5 add a new function that returns the underlying error.

Reply:
It would technically break the public API, but would not do so in a code-breaking way.

Since this function doesn't return anything, all existing code would look like

```
twirp.WriteError(resp, err)
```

This would not break if you started returning errors. i.e. you can get away with adding it in a minor version.

If that feels too close to the edge of the dark side.

> v5 add a new function that returns the underlying error.

💯 
I do not claim to know how to cleanly fix a public API. Probably something like this work?

- Add another method to return error
- Mark the current one deprecated
- In V6, drop the current method

Reply:
I like this discussion - let's move it to an issue though, since I don't think anybody else is seeing this :) @ofpiyush, can you make an issue?